### PR TITLE
[#4099] Prevent file_modified on non-write ops (4-1-stable)

### PR DESCRIPTION
**Pull request by alanking**
This change restricts when fileModified is called in rsModDataObjMeta. This should prevent replication (in the case of replication or compound hierarchies) on any non-write/create operation.

The replication resource has been modified such that the child replication list is generated at the time of fileModified instead of in resolve resource hierarchy. This ensures that the replication list is made available to the machine doing the replication as the property map may not be populated from machine to machine.

---
[CI tests passed](http://172.25.14.63:8080/job/test-ub14-oracle/1651/testReport/). Topo tests pending.

### CODE REVIEW

File: iRODS/server/config/lockFileDir/.gitignore [hunk 1]
```diff
@@ -1,2 +0,0 @@
-*
```
Comment (trel) [inline iRODS/server/config/lockFileDir/.gitignore]: was this removal intended?

File: plugins/resources/replication/librepl.cpp [hunk 2]
```diff
@@ -378,50 +416,177 @@ extern "C" {
...
+        ret = selected_hier_parser.str(current_resc_hier, name);
+        if (!ret.ok()) {
+            return PASSMSG(
+                       (boost::format(
+                        "[%s] - Failed to get hierarchy string for resc name [%s].") %
+                        __FUNCTION__ % name).str(), ret);
+        }
+        irods::hierarchy_parser current_hier_parser;
+        current_hier_parser.set_string(current_resc_hier);
+        //rodsLog(LOG_NOTICE, "[%s] - pared down hier to [%s]", __FUNCTION__, current_resc_hier.c_str());
```
Comment (trel) [inline plugins/resources/replication/librepl.cpp]: LOG_DEBUG?

File: iRODS/server/config/lockFileDir/.gitignore (hunk 1 above)
Comment (alanking) [reply iRODS/server/config/lockFileDir/.gitignore]: Nope, must have added it by accident. 4.1's build process (or at least run-in-place) seems to touch this file every time.

File: plugins/resources/replication/librepl.cpp [hunk 3]
```diff
@@ -378,50 +416,177 @@ extern "C" {
...
+        }
+        const std::string host_name(host_name_str);
+
+        // Loop over children and create a list of hierarchies to which objects will be replicated
+        float out_vote;
+        child_list_t repl_vector;
+        irods::error result = SUCCESS();
+
+        irods::resource_child_map::iterator it;
+        for ( it = _ctx.child_map().begin(); result.ok() && it != _ctx.child_map().end(); ++it ) {
```
Comment (korydraughn) [inline plugins/resources/replication/librepl.cpp]: Is `result.ok()` necessary in the condition?  It doesn't appear to change in the loop at all.

File: plugins/resources/replication/librepl.cpp [hunk 4]
```diff
@@ -215,22 +215,60 @@ extern "C" {
...
+            irods::file_object_ptr file_obj = boost::dynamic_pointer_cast<irods::file_object >( _ctx.fco() );
+            if (file_obj->l1_desc_idx() > 0) {
+                char* hier_str = getValByKey(&L1desc[file_obj->l1_desc_idx()].dataObjInp->condInput, RESC_HIER_STR_KW);
+                if (!hier_str) {
+                    return ERROR(SYS_INTERNAL_NULL_INPUT_ERR,
+                                 (boost::format(
+                                  "[%s] - No hierarchy string found in keywords for file object.") %
+                                  __FUNCTION__).str().c_str());
+                }
+                selected_parser.set_string(std::string(hier_str));
```
Comment (korydraughn) [inline plugins/resources/replication/librepl.cpp]: Is wrapping `hier_str` in `std::string` required?

File: plugins/resources/replication/librepl.cpp [hunk 5]
```diff
@@ -1369,47 +1446,36 @@ extern "C" {
...
-        // Make sure this is a valid repl operation.
-        if ( !( ret = replValidOperation( _ctx ) ).ok() ) {
-            return PASSMSG(
-                       (boost::format(
-                        "[%s] - Invalid operation on replicating resource.") %
-                        __FUNCTION__).str(), ret );
-        }
         // add ourselves to the hierarchy parser
-        else if ( !( ret = replAddSelfToHierarchy( _ctx, parser ) ).ok() ) {
+        irods::hierarchy_parser parser = *_inout_parser;
```
Comment (korydraughn) [inline plugins/resources/replication/librepl.cpp]: Would this be better if it were a reference or is it cheap to copy?

File: plugins/resources/replication/librepl.cpp (hunk 4 above)
Comment (alanking) [reply plugins/resources/replication/librepl.cpp]: Doesn't look like it. Removing!

File: plugins/resources/replication/librepl.cpp (hunk 3 above)
Comment (alanking) [reply plugins/resources/replication/librepl.cpp]: Good point. I believe I based this on the logic found in `replRedirectToChildren` wherein result is updated in the conditional block on 486. result is returned immediately after the loop. We should probably return the `PASSMSG` created there instead of just logging it.

File: plugins/resources/replication/librepl.cpp (hunk 5 above)
Comment (alanking) [reply plugins/resources/replication/librepl.cpp]: We can investigate that possibility for the 4.2+ implementation. Here, we are maintaining the existing behavior, and `resolve_resource_hierarchy` is not called frequently anyway. I am unsure of the ramifications of overwriting the value of the passed-in parser before calling `replSelectChild`.
